### PR TITLE
feat: add message to withdraw emission rewards

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

- Add a message to enable observers to collect emission rewards 

Closes: https://github.com/zeta-chain/node/issues/1653

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/emissions/abci.go [hunk 1]
```diff
@@ -146,3 +146,40 @@ func DistributeTssRewards(ctx sdk.Context, amount sdk.Int, bankKeeper types.Bank
...
+
+func EndBlocker(ctx sdk.Context, keeper keeper.Keeper) {
+	allWithdrawEmissions := keeper.GetAllWithdrawEmissions(ctx)
+	// delete all withdraw emissions irrespective of weather they are processed or not; this is to avoid processing the same withdraw emission again if in case it fails
+	// if a witdraw emission fails, it will need to be created again by the observer
+	defer func() {
+		for _, withdrawEmission := range allWithdrawEmissions {
+			keeper.DeleteWithdrawEmissions(ctx, withdrawEmission.Address)
+		}
+	}()
```
Comment (lumtis) [inline x/emissions/abci.go:158]: Why using a defer here? It seems we never return an error in the function body

**Review by lumtis — COMMENTED** [review]
I'm not sure to understand.
Why creating a withdraw object in the message and processing it in the endblocker, instead of directly processing it in the message?
What does the `WithdrawEmission` object brings on top of `WithdrawableEmission`?

**Comment by kingpinXD** [comment]
> I'm not sure to understand. Why creating a withdraw object in the message and processing it in the endblocker, instead of directly processing it in the message? What does the `WithdrawEmission` object brings on top of `WithdrawableEmission`?

This is to prevent more than one emission withdraw in the same block. Although I can't really think of a condition which,would allow observers to withdraw more rewards than they own , Using the endblock and processing all withdraws in the block at once makes the logic a bit safer 
I don't think it adds much of an overhead on top of a regular tx .

File: x/emissions/abci.go (hunk 1 above)
Comment (kingpinXD) [reply x/emissions/abci.go:158]: I used it just for code readability.

**Comment by lumtis** [comment]
> > I'm not sure to understand. Why creating a withdraw object in the message and processing it in the endblocker, instead of directly processing it in the message? What does the `WithdrawEmission` object brings on top of `WithdrawableEmission`?
> 
> This is to prevent more than one emission withdraw in the same block. Although I can't really think of a condition which,would allow observers to withdraw more rewards than they own , Using the endblock and processing all withdraws in the block at once makes the logic a bit safer I don't think it adds much of an overhead on top of a regular tx .

I don't get it. 

We have a `WithdrawableEmissions` object, already in the store, containing the amount an address can withdraw.
We're implementing a `MsgWithdrawEmission` that creates a `WithdrawEmissions` containing the exact same fields as above.
The end blocker iterates these `WithdrawEmissions`, send funds to user, delete `WithdrawEmissions` and subtract `WithdrawableEmissions` from value in `WithdrawEmissions`

Why not directly having `MsgWithdrawEmission` that decrease `WithdrawableEmissions` and send funds to user?

> This is to prevent more than one emission withdraw in the same block.

Why is it a problem if we verify that the withdrawn amount is below withdrawable amount?

> Although I can't really think of a condition which,would allow observers to withdraw more rewards than they own

How does using a intermediate end blocker add more protection for this?

**Comment by kingpinXD** [comment]
Closing pr for now to make changes mentioned above cc @lumtis

**Pull request closed without merging**